### ISSUE TITLE
src lint: warn if offsets (in refs, defs, docs, etc.) are char offsets, not byte offsets

### ISSUE THREAD

**Issue by sqs** [sourcegraph/srclib#110]

**Comment by bsummer4** [comment]
Currently, src produces incorrect results for srclib-haskell if it outputs byte offsets.

**Comment by sqs** [comment]
Yes, that's because it has that [hack](https://sourcegraph.com/sourcegraph.com/sourcegraph/srclib@master/.GoPackage/sourcegraph.com/sourcegraph/srclib/grapher/.def/NormalizeData#startline=111&endline=111) that converts all non-whitelisted source unit offsets from characters to bytes (which breaks if they're already bytes). Feel free to commit a change to that if statement.

**Comment by bsummer4** [comment]
The current behavior is actually much more convenient for srclib-haskell. I was hoping to delete the code that handles the conversion from char-offsets to bytes.

Could we make this opt-in, instead of having hard-coded special cases?

**Comment by sqs** [comment]
Yeah, making it opt-in seems good.

**Comment by bsummer4** [comment]
Do you think that setting {Srclibtoolchain.graph.OffsetType = "bytes" |
"characters"} would be an appropriate way to opt-in?

**Comment by sqs** [comment]
Yeah, that sounds good to me. Then I think we'll have to construct the Makefile recipes (that `src make` builds/uses) differently depending on the value of OffsetType, since right now all of the graphers pipe to `| src internal normalize-graph-data` to do the conversion. That `normalize-graph-data` subcommand has no easy way of knowing the source toolchain's config, but we do know that at Makefile creation time.

**Comment by bsummer4** [comment]
Ok, so: src reads the Srclibtoolchain file, creates a makefile, that makefile currently pipes the graph output to "normalize-graph-data", which currently looks at the unitType to see if it needs to convert the offsets. Is this correct?

And then, the change would be to change how the makefile is generated, so that normalize-graph-data is passed a flag?

This approach would avoid breaking things during transition. Pass a flag only if the field is set in the Srclibtoolchain, If no flag is passed, use the current hard-coded rules.

**Comment by sqs** [comment]
Sounds good!
